kafka.go

```diff
@@ -1,161 +1,215 @@
-	MetadataRetryBackoff time.Duration
+	DialTimeout  time.Duration
+	ReadTimeout  time.Duration
+	WriteTimeout time.Duration
 }
 
-// NewKafkaClient creates new client to publish event to kafka
-// The first producerConfiguration is the asynchronous producer configuration
-// meanwhile the second is the synchronous one
-func NewKafkaClient(realm string, brokerList []string, producerConfiguration ...*KafkaConfig) (*KafkaClient, error) {
-
-	var configAsync, configSync *sarama.Config
-	switch {
-	case len(producerConfiguration) > 1:
-		configAsync = configureKafka(producerConfiguration[0])
-		configSync = configureKafka(producerConfiguration[1])
-	case len(producerConfiguration) == 1:
-		configAsync = configureKafka(producerConfiguration[0])
-		configSync = sarama.NewConfig()
-	default:
-		configAsync = sarama.NewConfig()
-		configSync = sarama.NewConfig()
+func setConfig(writerConfig *kafka.WriterConfig, readerConfig *kafka.ReaderConfig, config *KafkaConfig) {
+
+	if config.ReadTimeout != 0 {
+		writerConfig.ReadTimeout = config.WriteTimeout
+	}
+	if config.WriteTimeout != 0 {
+		writerConfig.WriteTimeout = config.WriteTimeout
 	}
+	if config.DialTimeout != 0 {
+		dialer := &kafka.Dialer{
+			Timeout: config.DialTimeout,
+		}
+		writerConfig.Dialer = dialer
+		readerConfig.Dialer = dialer
+	}
+}
 
-	asyncProducer, err := sarama.NewAsyncProducer(brokerList, configAsync)
-	if err != nil {
-		return nil, err
+// NewKafkaClient create a new instance of KafkaClient
+func NewKafkaClient(brokers []string, realm string, config ...*KafkaConfig) (*KafkaClient, error) {
+
+	writerConfig := &kafka.WriterConfig{
+		Brokers:  brokers,
+		Balancer: &kafka.LeastBytes{},
 	}
 
-	configSync.Producer.Return.Successes = true
-	syncProducer, err := sarama.NewSyncProducer(brokerList, configSync)
-	if err != nil {
-		return nil, err
+	readerConfig := &kafka.ReaderConfig{
+		Brokers:  brokers,
+		MaxBytes: defaultReaderSize,
 	}
 
-	go listenToError(asyncProducer)
+	// set client configuration
+	if len(config) > 0 {
+		setConfig(writerConfig, readerConfig, config[0])
+	}
 
 	return &KafkaClient{
-		realm:         realm,
-		asyncProducer: asyncProducer,
-		syncProducer:  syncProducer,
+		realm:           realm,
+		publishConfig:   *writerConfig,
+		subscribeConfig: *readerConfig,
+		subscribeMap:    make(map[string]map[string]func(*Event, error)),
 	}, nil
 }
 
-func listenToError(producer sarama.AsyncProducer) {
-	for err := range producer.Errors() {
-		logrus.Error("unable to push event to Kafka stream: ", err)
+// Publish send event to message broker
+func (client *KafkaClient) Publish(publishBuilder *PublishBuilder) {
+	message, err := constructEvent(publishBuilder)
+	if err != nil {
+		logrus.Errorf("unable to construct event : %s , error : %v", publishBuilder.eventName, err)
+		return
 	}
-}
 
-type eventEntry struct {
-	Event
+	for _, topic := range publishBuilder.topic {
+		config := client.publishConfig
+		config.Topic = topic
+		writer := kafka.NewWriter(config)
```

Comment:
yes, we able to sent an event to multiple topics and the problem is, segmentio lib set the topic in config so we need to creating object every loop and every time we need to publish
but I think, we can make it non-blocking, so that would reduce the burden and also we can move line:88 before the loop. thanks

Reply:
what's the risk of storing the writers in memory so that you don't need to create-delete every time?

Reply:
creating writers need a topic and we get the topic by iterating publishBuilder.topic thats why we cant pool the writer in memory

Reply:
oh well